src/security/body_parse/body_parsing.cpp

```diff
@@ -0,0 +1,158 @@
+bool is_content_type(std::string_view actual, std::string_view tested) {
+  auto sv = actual;
+  while (sv.front() == ' ' || sv.front() == '\t') {
+    sv.remove_prefix(1);
+  }
+
+  if (actual.size() < tested.size()) {
+    return false;
+  }
+
+  for (std::size_t i = 0; i < tested.size(); i++) {
+    if (std::tolower(sv.at(i)) != tested.at(i)) {
+      return false;
+    }
+  }
+  sv.remove_prefix(tested.size());
+
+  if (!sv.empty()) {
+    auto next = sv.front();
+    if (next == ';' || next == ' ' || next == '\t') {
+      return true;
+    }
+    return false;
+  }
+
+  return true;
+}
+
+bool is_json(const ngx_http_request_t &req) {
+  const ngx_table_elt_t *ct = req.headers_in.content_type;
+  // don't look at ct->next; consider only the first value
+  return ct && is_content_type(datadog::nginx::to_string_view(ct->value),
+                               "application/json"sv);
+}
+
+bool is_multipart(const ngx_http_request_t &req) {
+  const ngx_table_elt_t *ct = req.headers_in.content_type;
+  return ct && is_content_type(datadog::nginx::to_string_view(ct->value),
+                               "multipart/form-data"sv);
+}
+
+bool is_urlencoded(const ngx_http_request_t &req) {
+  const ngx_table_elt_t *ct = req.headers_in.content_type;
+  return ct && is_content_type(datadog::nginx::to_string_view(ct->value),
+                               "application/x-www-form-urlencoded"sv);
+}
+
+}  // namespace
+
+namespace datadog::nginx::security {
+
+bool parse_body(ddwaf_obj &slot, ngx_http_request_t &req,
+                const ngx_chain_t &chain, std::size_t size,
+                DdwafMemres &memres) {
+  if (is_json(req)) {
+    // use rapidjson to parse:
+    bool success = parse_json(slot, req, chain, memres);
+    if (success) {
+      return true;
+    }
+  }
+
+  if (is_multipart(req)) {
+    std::optional<HttpContentType> ct = HttpContentType::for_string(
+        to_string_view(req.headers_in.content_type->value));
+    if (!ct) {
+      ngx_log_error(NGX_LOG_NOTICE, req.connection->log, 0,
+                    "multipart: invalid multipart/form-data content-type");
+      return false;
+    }
+
+    return parse_multipart(slot, req, *ct, chain, memres);
+  }
+
+  // linearize the input
+  char *buf = memres.allocate_string(size);
+  NgxChainInputStream stream{&chain};
+  auto read = stream.read(reinterpret_cast<std::uint8_t *>(buf), size);
+  if (read < size) {
+    throw std::runtime_error(
+        "mismatch between declared size and read size (read is smaller than "
+        "declared)");
+  }
+
+  if (is_urlencoded(req)) {
+    QueryStringIter it{
+        {buf, size}, memres, '&', QueryStringIter::trim_mode::no_trim};
+
+    // count key occurrences
+    std::unordered_map<std::string_view, std::size_t> bag;
+    for (; !it.ended(); ++it) {
+      std::string_view cur_key = it.cur_key();
+      bag[cur_key]++;
+    }
+
+    // allocate all ddwaf_obj, set keys
+    std::unordered_map<std::string_view, ddwaf_obj *> key_index;
```

Comment:
You could use the same map for both key counting and keeping track of duplicate keys, just replace the value with `std::pair<std::size_t, ddwaf_obj *>` or, better yet, a struct.

Reply:
It might be more confusing but it avoids the generation of the same hash table twice.

Reply:
ok i put a union in place